### PR TITLE
refs #1439 - adding methods setFieldValue, setFieldValueName, setFiel…

### PR DESCRIPTION
**Pull request by Macek007**
…dValueLabel

Methods sets form field value by CSS3, XPath, name or label. Form reference is not needed.

SlimerJS ends me with errors - same as origin/master - so I ignered them.

Documentation is missing, please look first to my code.

Thanks

### CODE REVIEW

**Comment by mickaelandrieu** [comment]
> SlimerJS ends me with errors - same as origin/master - so I ignered them.

Seems to be good now :)

Can be useful :+1: 

Please add docs.

**Comment by mickaelandrieu** [comment]
@istr my bad ! didnt saw you didnt agree with this feature and I've already merged docs ... go or no go ?

**Comment by istr** [comment]
@mickaelandrieu No problem at all. I think this feature is very useful; the only thing I know @n1k0 might definitely raise an eyebrow about is the ```switch``` construct in the code. I would tend to merge it, but give me a chance for a thorough review of the code this evening.

**Comment by mickaelandrieu** [comment]
:+1:  as you like ;)

File: modules/casper.js [hunk 1]
```diff
@@ -1686,6 +1686,62 @@ Casper.prototype.setContent = function setContent(content) {
     return this;
 };
 
+
+/**
+ * Sets a value to form field by CSS3, XPath selector or by it's name attribute or label text.
+ *
```
Comment (istr) [inline modules/casper.js]: "...or by its name..." (its -> pronoun, it's -> it is / it has)

File: modules/clientutils.js [hunk 2]
```diff
@@ -269,30 +271,13 @@
...
-                },
-                names: function(elementName, formSelector) {
-                    return this.findAll('[name="' + elementName + '"]', form);
-                },
-                xpath: function(xpath, formSelector) {
-                    return this.findAll({type: "xpath", path: xpath}, form);
-                }
-            };
-
             for (var fieldSelector in vals) {
```
Comment (istr) [inline modules/clientutils.js]: What is the reason to replace this ```map```able construct with a switch? (see below)

File: modules/clientutils.js [hunk 3]
```diff
@@ -659,6 +644,49 @@
...
+                selector = selector.path;
+            }
+
+            switch (type) {
+                case 'css': // do nothing
+                    ret = selector;
+                    break;
+                case 'name': // convert to css 
+                case 'names': 
+                    ret = '[name="' + selector + '"]';
```
Comment (istr) [inline modules/clientutils.js]: If the only reason for the switch is the singular alias, you might as well add it in the object-style as seen in the existing code, i.e.

```js
var finders = {
...
labels: ...
};
finders.label = finders.labels;
```

File: tests/suites/casper/formfill.js [hunk 4]
```diff
@@ -1,5 +1,6 @@
 /*eslint strict:0*/
 var fs = require('fs');
+var x = require('casper').selectXPath;
```
Comment (istr) [inline tests/suites/casper/formfill.js]: Please try to avoid single letter variable names, even in test suites.

**Comment by istr** [comment]
@mickaelandrieu Apart from some very minor questions / remarks I am ok to merge this. You decide.

**Comment by mickaelandrieu** [comment]
Yes, I'm still convinced by the benefit of this contribution.

Can you address the comments @Macek007 ?

Thank you

File: modules/clientutils.js (hunk 3 above)
Comment (Macek007) [inline modules/clientutils.js]: The reason was that fill() did a lot of things and I needed similar functionality for my methods - that's why I isolated of making selector (makeSelector method). 
"Finders" needed more parameters, in all cases is findAll() called etc. On other hand makeSelector() is simple method with clear interface.  
I think using switch in this case is OK but I'll would rewrite it if you want.
(btw. I think in method setField() is bad example of using mega switch with a lot of logic)

File: tests/suites/casper/formfill.js (hunk 4 above)
Comment (Macek007) [inline tests/suites/casper/formfill.js]: I thought that it is recommended pattern for XPath expressions according this example http://docs.casperjs.org/en/latest/selectors.html#index-2

File: modules/clientutils.js (hunk 3 above)
Comment (istr) [inline modules/clientutils.js]: Totally ok to factor out the ```var finders = {...``` logic into ```makeSelector``` if you need to re-use it. But I would prefer to stick with the existing "object switch" method among other things to keep the change-set small.

File: tests/suites/casper/formfill.js (hunk 4 above)
Comment (istr) [inline tests/suites/casper/formfill.js]: Ok, never mind. :-) This is just a personal preference (it is easier to e.g. ```grep``` for variables with more than one letter). You decide.

**Comment by mickaelandrieu** [comment]
@istr can I assign myself on it ? should be merged as the docs are already online (http://docs.casperjs.org/en/latest/modules/clientutils.html?highlight=setfieldvalue#setfieldvalue)

Should I remove the docs now, to reflect the real API ?

**Comment by istr** [comment]
> @istr can I assign myself on it ? should be merged as the docs are already online

Yes, of course. Maybe just clone the PR and rebase it, if @Macek007 does not react on this ping?

**Comment by jefleponot** [comment]
I can do it if you want @mickaelandrieu

**Comment by mickaelandrieu** [comment]
@jefleponot go for it (thanks ! Realy sorry I'm so busy :/)

**Comment by Macek007** [comment]
Hi, sorry I couldn't react earlier to this PR... it seems that @jefleponot has finished it. Do you want something else with this PR?

**Comment by mickaelandrieu** [comment]
@Macek007 no, thank you for your contribution, shipped with 1.1 ;)

**Pull request closed without merging**